diamond-api/src/render.rs

```diff
@@ -144,54 +152,252 @@ fn walk(dir: &Path, metric: &str, interval: Interval) -> Result<Vec<RenderPoint>
-pub struct RenderResponceEntry {
+pub struct RenderResponseEntry {
     target: String,
     datapoints: Vec<RenderPoint>,
 }
 
 #[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
-pub struct RenderResponce {
-    entries: Vec<Option<RenderResponceEntry>>,
+pub struct RenderResponse {
+    entries: Vec<Option<RenderResponseEntry>>,
 }
 
 pub fn render_handler(
-    state: Data<Args>,
-    params: RenderQuery,
+    ctx: Data<Context>,
+    query: RenderQuery,
 ) -> impl Future<Item = HttpResponse, Error = failure::Error> {
-    let dir = &state.path;
-
-    match Interval::new(params.from, params.until).map_err(err_msg) {
+    match Interval::new(query.from, query.until).map_err(err_msg) {
         Ok(interval) => {
-            let response: Result<Vec<RenderResponceEntry>, ResponseError> = params
+            let response: Result<Vec<RenderResponseEntry>, ResponseError> = query
                 .target
                 .into_iter()
-                .map(|x| {
-                    walk(&dir, &x, interval).map(|datapoints| RenderResponceEntry {
-                        datapoints,
-                        target: x,
+                .map(|metric| {
+                    let w = walker(ctx.walker.clone());
+                    w(&metric, interval).map(|points| RenderResponseEntry {
+                        datapoints: points,
+                        target: metric,
                     })
                 })
                 .collect();
 
             match response {
-                Ok(response) => result(Ok(HttpResponse::Ok().json(response))),
+                Ok(response) => result(Ok(format_response(response, query.format))),
                 Err(e) => err(e.into()),
             }
         }
         Err(e) => err(e),
     }
 }
 
+// TODO: Extract code BELOW to response_format_csv module
+trait IntoCsv {
+    fn into_csv(self) -> String;
+}
+
+impl<T: IntoCsv> IntoCsv for Vec<T> {
+    fn into_csv(self) -> String {
+        self.into_iter()
+            .map(|item| item.into_csv())
+            .collect::<Vec<String>>()
+            .join("\n") + "\n"
+    }
+}
+
+impl IntoCsv for RenderResponseEntry {
+    fn into_csv(self) -> String {
+        let metric = self.target;
+        let lines: Vec<String> = self
+            .datapoints
+            .into_iter()
+            .map(|point| {
+                let (ts, val) = (point.1, point.0.unwrap_or(0.0 as f64));
+                // TODO: Use `csv` crate instead of "manual" string formatting
+                format!("{},{},{}", metric, ts, val)
+            })
+            .collect();
+        lines.join("\n")
+    }
+}
+// TODO: Extract code ABOVE to response_format_csv module
+
+fn format_response(response: Vec<RenderResponseEntry>, format: RenderFormat) -> HttpResponse {
+    match format {
+        RenderFormat::Json => HttpResponse::Ok().json(response),
+        RenderFormat::Csv => HttpResponse::Ok().body(response.into_csv()),
```

Comment:
`Content-Type` `text/csv` header should be added as well

Reply:
Done